pkg/plugin/plugin.go

```diff
@@ -27,50 +25,28 @@ func (p *Plugin) ValidateInternal(new, old *acsapi.ContainerService) []error {
 	return validate.ContainerService(new, old)
 }
 
-func (p *Plugin) GenerateConfig(cs *acsapi.ContainerService, configBytes []byte) ([]byte, error) {
-	var c *config.Config
-	if len(configBytes) > 0 {
-		err := yaml.Unmarshal(configBytes, &c)
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		c = &config.Config{}
-	}
-
-	err := config.Upgrade(cs, c)
-	if err != nil {
-		return nil, err
+func (p *Plugin) GenerateConfig(cs *acsapi.ContainerService) error {
+	// TODO should we save off the original config here and if there are any errors we can restore it?
```

Comment:
I think we shouldn't be mutating stored config if we don't apply it in a cluster. Not sure what exactly does that mean for this case.

Reply:
+1, that's what I was thinking with this comment.  We should either return a new version of the config and expect the orchestrator to handle adding it to the struct or we should retain a copy and revert it if we return an error.

Since `GenerateConfig` calls both `Upgrade` and `Generate` I was leaning towards just preserving the config internally but would be open to other solutions

Reply:
then again, declarative is exactly that....mutating desired state prior to applying it :-)  That begs the question of versioned storage of the internal model.  Maybe @amanohar can comment on how a rollback might work in that case or if it is forever forward moving